modules/dag-l1/src/main/scala/org/tessellation/dag/l1/domain/block/BlockService.scala

```diff
@@ -62,35 +58,29 @@ object BlockService {
         } yield ()
 
       private def prepareBalances(transactions: Set[Transaction]): F[Map[Address, Balance]] = {
-        val sources = transactions.groupBy(_.source)
-        val destinations = transactions.groupBy(_.destination)
-        val all = sources.combine(destinations)
-        val addressBalanceChanges = all.map {
-          case (address, txs) =>
-            val change = txs.foldLeft(BigInt(0L))(
-              (change, tx) =>
-                tx match {
-                  case Transaction(`address`, _, amount, fee, _, _) => change - amount.coerce - fee.coerce
-                  case Transaction(_, `address`, amount, _, _, _)   => change + amount.coerce
-                  case _                                            => change
-                }
-            )
+        val sources = transactions.groupBy(_.source).keySet
+        val destinations = transactions.groupBy(_.destination).keySet
 
-            address -> change
-        }
+        val addresses = sources ++ destinations
 
-        addressBalanceChanges.toList.traverse {
-          case (address, balanceChange) =>
-            for {
-              current <- addressStorage.getBalance(address)
-              newBalance <- refineV[NonNegative]
-                .apply(current.coerce + balanceChange)
-                .bimap(BalanceNegative, Balance(_))
-                .liftTo[F]
-            } yield (address, newBalance)
+        def applyTransactions(balance: Balance, address: Address): Either[BalanceOutOfRange, Balance] =
+          transactions.foldLeft(balance.asRight[BalanceOutOfRange]) { (acc, tx) =>
```

Comment:
You are going through all the transactions just to extract balances for single addresses. Maybe there is something like `foldF` where you could traverse `Map[Address, Transaction]` and fold transactions with initial balance for every address?

Reply:
I can use `filter` upfront, but I don't see any optimization in that. anyway I'd need to iterate through all transactions